File: app/src/main/java/skripsi/kohonen/signaturerecognition/Kohonen.java
package skripsi.kohonen.signaturerecognition;

import android.util.Log;

import java.util.ArrayList;

public class Kohonen {

    private final int panjangPola = 40000;
    private final double alpha;
    private double[] d;
    private int panjangX;
    MainActivity mainActivity;
    private final ArrayList<byte[]> x;
    public final ArrayList<String> huruf;
    public final ArrayList<byte[]> nilai;
    public final ArrayList<String> nama;

    private final Database db;

    public Kohonen(Database db) {
        alpha = 0.6;
        panjangX = 1;
        this.db = db;
        d = new double[panjangX];
        huruf = new ArrayList<>();
        x = new ArrayList<>();
        nama = new ArrayList<>();
        nilai = new ArrayList<>();
        isiX();
    }

    private void isiX() {
        x.clear();
        huruf.clear();
        ArrayList<TandaTangan> listttd = db.getAllttd();
        for (int i = 0; i < db.getttdsCount(); i++) {
            TandaTangan kat = listttd.get(i);
            byte[] bt = TandaTangan.strToArr(kat.getNilai());
            x.add(bt);
            String str = kat.getnama();
            huruf.add(str);
        }
        panjangX = x.size();
    }

    public static double asli = 0;
    public static double minimal = 0;

    public static int nilaiMaxJarak = 0;
    public static double nilaiJarak = 0;
    public  static  String name ="";
    private void hitungX(byte[] arr) {
        d = new double[panjangX];
        double minjarak = 0;
        double maxJarak = 0;

        clearArray(d);
        for (int i = 0; i < panjangX; i++) {
            for (int j = 0; j < panjangPola; j++) {
                d[i] += Math.pow((x.get(i)[j] - arr[j]), 2);
            }
            if (i == 0) {
                minjarak = d[0];
            } else if ((i > 0) && (d[i] < minjarak)) {
                minjarak = d[i];
            }
/*
            if (i == 0 ) {
                maxJarak = d[0];
            } else if ((i > 0) && (d[i] > maxJarak)) {
                maxJarak = d[i];
            }*/

            Log.d("jarak cluster:" + (i), String.valueOf(d[i]));
        }

       /* nilaiMaxJarak = (int) maxJarak;*/
        Log.d("jarak terkecil: ", String.valueOf(minjarak));
        minimal = minjarak;

       /* asli = (20000 - minjarak) / (20000 - 0) * 100;
        Log.d("nilai", String.valueOf(nilai));*/
    }


    private void clearArray(double[] nodeArray) {
        for (int i = 0; i < panjangX; i++) {
            nodeArray[i] = 0.0;
        }
    }

    private int minimum(ArrayList<Double> nodeArray) {
        double min = nodeArray.get(0);
        int index = 0;
        for (int i = 0; i < nodeArray.size(); i++) {
            if (nodeArray.get(i) < min) {
                min = nodeArray.get(i);
                index = i;
            }
        }
        return index;
    }

    public ArrayList<String> tesDataBaru(byte[] arr) {
        ArrayList<String> arrStr = new ArrayList<>();
        hitungX(arr);
        ArrayList<Double> arrD = new ArrayList<>();
        ArrayList<String> arrS = new ArrayList<>();
        for (int x = 0; x < huruf.size(); x++) {
            arrD.add(d[x]);
            arrS.add(huruf.get(x));
        }
        int min = minimum(arrD);
        arrStr.add(huruf.get(min));
        name = huruf.get(min);
        Log.d("nama: ", String.valueOf((huruf.get(min))));
        return arrStr;
    }
    public ArrayList<Double> tesBerdasarkanNama(byte[][] arrBiner) {
        ArrayList<Double> arrHasil = new ArrayList<>();
        double[] jarak = new double[arrBiner.length];
        for (int i = 0; i < arrBiner.length; i++) {
            for (int j = 0; j < panjangPola; j++) {
                jarak[i] += Math.pow((arrBiner[i][j] - arrBiner[0][j]), 2);
            }
            arrHasil.add(jarak[i]);
        }
        return arrHasil;
    }

    ArrayList<Integer> sama(ArrayList<String> arrS, String str) {
        ArrayList<Integer> arrInt = new ArrayList<>();
        for (int i = 0; i < arrS.size(); i++) {
            if (arrS.get(i).equals(str)) {
                arrInt.add(i);
            }
        }
        return arrInt;
    }

    public void hapus(int pos) {
        huruf.remove(pos);
        x.remove(pos);
        panjangX = x.size();
    }

    public void ubah(int pos, String str) {
        huruf.set(pos, str);
    }

    public void hapusSemua() {
        x.clear();
        huruf.clear();
    }

    public void latihDataBaru(byte[] arr, String str) {
        byte[] bt = new byte[arr.length];
        System.arraycopy(arr, 0, bt, 0, arr.length);
        x.add(bt);
        huruf.add(str);
        panjangX = x.size();
    }
}

File: app/src/main/java/skripsi/kohonen/signaturerecognition/Kohonen.kt
package skripsi.kohonen.signaturerecognition

import android.util.Log
import java.util.*

class Kohonen(private val db: Database?) {
    private val panjangPola = 40000
    private val alpha = 0.6
    private var d: DoubleArray
    private var panjangX = 1
    var mainActivity: MainActivity? = null
    private val x: ArrayList<ByteArray>
    val huruf: ArrayList<String?>
    val nilai: ArrayList<ByteArray>
    val nama: ArrayList<String>
    private fun isiX() {
        x.clear()
        huruf.clear()
        val listttd = db!!.allttd
        for (i in 0 until db.getttdsCount()) {
            val kat = listttd!![i]
            val bt: ByteArray = TandaTangan.Companion.strToArr(kat!!.nilai)
            x.add(bt)
            val str = kat!!.getnama()
            huruf.add(str)
        }
        panjangX = x.size
    }

    private fun hitungX(arr: ByteArray) {
        d = DoubleArray(panjangX)
        var minjarak = 0.0
        val maxJarak = 0.0
        clearArray(d)
        for (i in 0 until panjangX) {
            for (j in 0 until panjangPola) {
                d[i] += Math.pow((x[i][j] - arr[j]).toDouble(), 2.0)
            }
            if (i == 0) {
                minjarak = d[0]
            } else if (i > 0 && d[i] < minjarak) {
                minjarak = d[i]
            }
            /*
            if (i == 0 ) {
                maxJarak = d[0];
            } else if ((i > 0) && (d[i] > maxJarak)) {
                maxJarak = d[i];
            }*/Log.d("jarak cluster:$i", d[i].toString())
        }

        /* nilaiMaxJarak = (int) maxJarak;*/Log.d("jarak terkecil: ", minjarak.toString())
        minimal = minjarak

        /* asli = (20000 - minjarak) / (20000 - 0) * 100;
        Log.d("nilai", String.valueOf(nilai));*/
    }

    private fun clearArray(nodeArray: DoubleArray) {
        for (i in 0 until panjangX) {
            nodeArray[i] = 0.0
        }
    }

    private fun minimum(nodeArray: ArrayList<Double>): Int {
        var min = nodeArray[0]
        var index = 0
        for (i in nodeArray.indices) {
            if (nodeArray[i] < min) {
                min = nodeArray[i]
                index = i
            }
        }
        return index
    }

    fun tesDataBaru(arr: ByteArray): ArrayList<String?> {
        val arrStr = ArrayList<String?>()
        hitungX(arr)
        val arrD = ArrayList<Double>()
        val arrS = ArrayList<String?>()
        for (x in huruf.indices) {
            arrD.add(d[x])
            arrS.add(huruf[x])
        }
        val min = minimum(arrD)
        arrStr.add(huruf[min])
        name = huruf[min]
        Log.d("nama: ", huruf[min].toString())
        return arrStr
    }

    fun tesBerdasarkanNama(arrBiner: Array<ByteArray>): ArrayList<Double> {
        val arrHasil = ArrayList<Double>()
        val jarak = DoubleArray(arrBiner.size)
        for (i in arrBiner.indices) {
            for (j in 0 until panjangPola) {
                jarak[i] += Math.pow((arrBiner[i][j] - arrBiner[0][j]).toDouble(), 2.0)
            }
            arrHasil.add(jarak[i])
        }
        return arrHasil
    }

    fun sama(arrS: ArrayList<String>, str: String): ArrayList<Int> {
        val arrInt = ArrayList<Int>()
        for (i in arrS.indices) {
            if (arrS[i] == str) {
                arrInt.add(i)
            }
        }
        return arrInt
    }

    fun hapus(pos: Int) {
        huruf.removeAt(pos)
        x.removeAt(pos)
        panjangX = x.size
    }

    fun ubah(pos: Int, str: String?) {
        huruf[pos] = str
    }

    fun hapusSemua() {
        x.clear()
        huruf.clear()
    }

    fun latihDataBaru(arr: ByteArray, str: String?) {
        val bt = ByteArray(arr.size)
        System.arraycopy(arr, 0, bt, 0, arr.size)
        x.add(bt)
        huruf.add(str)
        panjangX = x.size
    }

    companion object {
        var asli = 0.0
        var minimal = 0.0
        var nilaiMaxJarak = 0
        var nilaiJarak = 0.0
        var name: String? = ""
    }

    init {
        d = DoubleArray(panjangX)
        huruf = ArrayList()
        x = ArrayList()
        nama = ArrayList()
        nilai = ArrayList()
        isiX()
    }
}